daemons/gptp/common/ieee1588clock.cpp

```diff
@@ -350,26 +358,44 @@ void IEEE1588Clock::setMasterOffset
 	}
 
 	if( _syntonize ) {
-		if( _new_syntonization_set_point ) {
+		if( _new_syntonization_set_point || _phase_error_violation > PHASE_ERROR_MAX_COUNT ) {
 			_new_syntonization_set_point = false;
+			_phase_error_violation = 0;
 			if( _timestamper ) {
 				/* Make sure that there are no transmit operations
 				   in progress */
 				getTxLockAll();
+				if (port->getTestMode()) {
+					GPTP_LOG_STATUS("Adjust clock phase offset:%lld", -master_local_offset);
+				}
 				_timestamper->HWTimestamper_adjclockphase
 					( -master_local_offset );
 				_master_local_freq_offset_init = false;
+				restartPDelayAll();
 				putTxLockAll();
 				master_local_offset = 0;
 			}
 		}
+
 		// Adjust for frequency offset
 		long double phase_error = (long double) -master_local_offset;
-		_ppm += (float) (INTEGRAL*phase_error +
-			 PROPORTIONAL*((master_local_freq_offset-1.0)*1000000));
+		if( fabsl(phase_error) > PHASE_ERROR_THRESHOLD ) {
+			++_phase_error_violation;
+		} else {
+			_phase_error_violation = 0;
+
+			float syncPerSec = (float)(1.0 / pow((float)2, port->getSyncInterval()));
```

Comment:
Maybe we should add a getSyncPerSec() method to the port? Seems like overkill to call a pow() operation every time this runs - although I realize it doesn't run that often. I don't consider this a "blocking" comment - it's just a suggestion.